keyvalue/clients/src/main/java/tachyon/client/keyvalue/BaseKeyValueStoreWriter.java

```diff
@@ -0,0 +1,174 @@
+
+import com.google.common.base.Preconditions;
+
+import tachyon.Constants;
+import tachyon.TachyonURI;
+import tachyon.client.ClientContext;
+import tachyon.client.file.TachyonFile;
+import tachyon.client.file.TachyonFileSystem;
+import tachyon.client.file.TachyonFileSystem.TachyonFileSystemFactory;
+import tachyon.conf.TachyonConf;
+import tachyon.exception.TachyonException;
+import tachyon.thrift.PartitionInfo;
+import tachyon.util.io.BufferUtils;
+
+/**
+ * Default implementation of {@link KeyValueStoreWriter} to create a Tachyon key-value store.
+ */
+@NotThreadSafe
+class BaseKeyValueStoreWriter implements KeyValueStoreWriter {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+
+  private final TachyonFileSystem mTfs = TachyonFileSystemFactory.get();
+  private final TachyonConf mConf = ClientContext.getConf();
+  private final InetSocketAddress mMasterAddress = ClientContext.getMasterAddress();
+  private final KeyValueMasterClient mMasterClient;
+  private final TachyonURI mStoreUri;
+
+  private long mPartitionIndex;
+  private KeyValuePartitionWriter mWriter = null;
+  /** min key in the current partition */
+  private ByteBuffer mKeyStart = null;
+  /** max key in the current partition */
+  private ByteBuffer mKeyLimit = null;
+  /** whether this writer is closed */
+  private boolean mClosed;
+  /** whether this writer is canceled */
+  private boolean mCanceled;
+
+  /**
+   * Constructs a {@link BaseKeyValueStoreWriter}. This constructor will create a new key-value
+   * store at the given {@link TachyonURI}.
+   *
+   * @param uri URI of the store
+   * @throws IOException if a non-Tachyon exception occurs
+   * @throws TachyonException if an unexpected Tachyon exception is thrown
+   */
+  BaseKeyValueStoreWriter(TachyonURI uri) throws IOException, TachyonException {
+    LOG.info("Create KeyValueStoreWriter for {}", uri);
+    mMasterClient = new KeyValueMasterClient(mMasterAddress, mConf);
+
+    mStoreUri = Preconditions.checkNotNull(uri);
+    mMasterClient.createStore(mStoreUri);
+    mPartitionIndex = 0;
+    mClosed = false;
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (mClosed) {
+      return;
+    }
+    try {
+      if (mCanceled) {
+        mWriter.cancel();
+        // TODO(binfan): cancel all other written partitions
+      } else {
+        completePartition();
+        mMasterClient.completeStore(mStoreUri);
+      }
+    } catch (TachyonException e) {
+      throw new IOException(e);
+    } finally {
+      mMasterClient.close();
+    }
+    mClosed = true;
+  }
+
+  @Override
+  public void cancel() throws IOException {
+    mCanceled = true;
+    close();
+  }
+
+  @Override
+  public void put(byte[] key, byte[] value) throws IOException, TachyonException {
+    Preconditions.checkNotNull(key, "Cannot put a null key");
+    Preconditions.checkNotNull(value, "Cannot put a null value");
+    if (mWriter == null || mWriter.isFull()) {
```

Comment:
If mWriter is not full, but the remaining space is not enough to hold the key and value, shouldn't the current partition be completed and a new partition be created?